File: 90-subsets-ii/subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:

        def backtrack(i,subset):
            if i == len(nums):
                ans.append(subset.copy())
                return

            subset.append(nums[i])
            backtrack(i+1,subset)

            subset.pop()

            while i+1 < len(nums) and nums[i] == nums[i+1]:
                i += 1

            backtrack(i+1,subset)

        ans = []
        nums.sort()
        backtrack(0,[])
        return ans


        

File: 90-subsets-ii/subsets-ii.php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function subsetsWithDup($nums) {
        sort($nums);
        $this->ans = [];
        $this->nums = $nums;
        $this->backtrack([],0);
        return $this->ans;
    }

    function backtrack($cur,$index){
        $this->ans[] = $cur;

        for($i=$index;$i<count($this->nums);$i++){
            if($i > $index && $this->nums[$i] == $this->nums[$i-1]){
                continue;
            }
            $cur[] = $this->nums[$i];
            $this->backtrack($cur,$i+1);
            array_pop($cur);
        }
    }
}